src/deepMap.ts

```diff
@@ -1,65 +1,85 @@
-    private root: Map<any, any>
-    private closest: Map<any, any>
-    private closestIdx: number = 0
     isDisposed = false
+    private hash: string
 
-    constructor(private base: Map<any, any>, private args: any[]) {
-        let current: undefined | Map<any, any> = (this.closest = this.root = base)
-        let i = 0
-        for (; i < this.args.length - 1; i++) {
-            current = current!.get(args[i])
-            if (current) this.closest = current
-            else break
+    constructor(
+        private readonly state: {
+            store: Map<string, T>
+            weakHashes: WeakMap<object | Function | Symbol, string>
+            strongHashes: Map<Symbol, string>
+            hashId: number
+        },
+        args: Array<Readonly<unknown>>
+    ) {
+        this.hash = args.map((arg) => this.prehash(arg)).join("")
+    }
+
+    private prehash(arg: Readonly<unknown>): string | number {
+        if (typeof arg === "string") {
+            return `s:${arg}`
+        }
+        if (typeof arg === "number") {
+            return `n:${arg}`
+        }
+        if (arg === null) {
+            return "null"
+        }
+        if (typeof arg === "boolean") {
+            return arg ? "true" : "false"
         }
-        this.closestIdx = i
+        if (typeof arg === "undefined") {
+            return "undefined"
+        }
+        if (typeof arg === "object" || typeof arg === "function" || typeof arg === "symbol") {
+            let hashes = this.state.weakHashes
+            if (typeof arg === "symbol" && !supportedInWeakMap(arg)) {
+                hashes = this.state.strongHashes
+            }
+
+            if (hashes.has(arg)) return hashes.get(arg)!
+
+            const hash = `o:${this.state.hashId++}`
+            hashes.set(arg, hash)
+            return hash
+        }
+        if (typeof arg === "bigint") {
+            return `N:${arg}`
+        }
+
+        throw new Error("Unknown type")
     }
 
     exists(): boolean {
         this.assertNotDisposed()
-        const l = this.args.length
-        return this.closestIdx >= l - 1 && this.closest.has(this.args[l - 1])
+        return this.state.store.has(this.hash)
     }
 
     get(): T {
         this.assertNotDisposed()
         if (!this.exists()) throw new Error("Entry doesn't exist")
-        return this.closest.get(this.args[this.args.length - 1])
+        return this.state.store.get(this.hash)!
     }
 
     set(value: T) {
         this.assertNotDisposed()
-        const l = this.args.length
-        let current: Map<any, any> = this.closest
-        // create remaining maps
-        for (let i = this.closestIdx; i < l - 1; i++) {
-            const m = new Map()
-            current.set(this.args[i], m)
-            current = m
-        }
-        this.closestIdx = l - 1
-        this.closest = current
-        current.set(this.args[l - 1], value)
+        this.state.store.set(this.hash, value)
     }
 
     delete() {
```

Comment:
I think we have to leak them unfortunately.

Registered symbols are never collected, hence not allowed in a WeakMap, hence no way to trigger something on their disposal. I'm not too worried about those as I guess they are very rare.

Regular symbols on Firefox is a bigger deal. We can't just delete from `this.state.strongHashes` in `delete()` though, because the same symbol hash might have been used by multiple cached calls to our function. We would have to store a reference count in strongHashes and update it.

I don't see any technical blocker, but not sure if it's worth it just for Firefox